sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/expressions/Transform.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2.expressions;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * Represents a transform function in the public logical expression API.
+ * <p>
+ * For example, the transform date(ts) is used to derive a date value from a timestamp column. The
+ * transform name is "date" and its argument is a reference to the "ts" column.
+ */
+@Experimental
+public interface Transform extends Expression {
+  /**
+   * Returns the transform function name.
+   */
+  String name();
```

Comment:
how would a library developer know what names are out there?

Reply:
There isn't a restriction on names. Using the current implementation, it would be possible to express a custom partition function like this: `PARTITION BY (range(last_name, 'A', 'C', 'G', ...))`.

Spark defines commonly used transformations like bucketing and date transforms. Those transforms will be well-defined so they can eventually be used to pass values to Spark for metadata-only queries.

Reply:
Not sure if I asked the question clearly. Say I'm a developer that wants to built a new data source, will I need to know what the possible transform functions are? Or are those completely hidden from me?

Reply:
When it catalog API is added, it will link to the `Expressions` class that includes known transforms that have meaning to Spark. It will also note that any transform can be expressed, like I said above. The `range` transform example can be parsed and passed to the source, it just isn't one defined by Spark.